### PR TITLE
Add constraint filter field to /images/create

### PR DESCRIPTION
**Pull request by kyhavlov**
This PR adds the ability to pass a comma separated list of constraints in order to specify which hosts should pull or delete the image.

I'm not sure what the convention is around adding features like this that are only applicable to swarm and not docker engine itself, so I wrote it in the simplest way possible for now. If there's a different way this should be done, please comment.

Addresses #1692.

### CODE REVIEW

**Comment by allencloud** [comment]
@kyhavlov  Thanks for your work. First, can you squash two commits into one?

**Comment by kyhavlov** [comment]
@allencloud done

File: cluster/swarm/cluster.go [hunk 1]
```diff
@@ -574,12 +575,20 @@ func (c *Cluster) RemoveVolumes(name string) (bool, error) {
 }
 
 // Pull is exported
-func (c *Cluster) Pull(name string, authConfig *types.AuthConfig, callback func(where, status string, err error)) {
+func (c *Cluster) Pull(name string, constraints []string, authConfig *types.AuthConfig, callback func(where, status string, err error)) {
 	var wg sync.WaitGroup
 
 	c.RLock()
-	for _, e := range c.engines {
+
```
Comment (allencloud) [inline cluster/swarm/cluster.go]: I think this part does not have to be after `c.RLock()`. It will abuse the read lock here. @kyhavlov

**Comment by kyhavlov** [comment]
Addressed PR comment and added constraints field for deleting images as well

**Comment by IsmailMehdi** [comment]
👍  This is a nice way to improve the swarm's use of images. It would be critical in large clusters with various images.

**Comment by kyhavlov** [comment]
can any of the maintainers comment on what I should do to move this along? 
@nishanttotla @dongluochen

**Comment by chanwit** [comment]
@kyhavlov This is awesome! I'd love to help you to get this through.
  - Please clarify it's a bit, like putting an example of how people use this in the description of this PR.

In my #19785, I proposed a use of `-e` for the `pull` command. But it requires a change on the `docker` client.

Maybe your PR is a good starting point on implementing something like:

`docker pull nginx@constraint:node==node1,constraint:node!=node1`.

The `docker` client is already accepting this syntax. You can just week your parsing function to parse a specific `@` tag of the image then perform similar thing.

  - And right after we're review your design and move on to the code review, please also have an integration test for your changes as a `.bats` file (reside in the `test` directory).

Keep on hacking @kyhavlov, I'll try my best to help this PR gets through!!

**Comment by kyhavlov** [comment]
@chanwit:  `@constraint:node==node1` can't be the tag itself though, can it? What about when we want to specify a tag and constraints?

**Comment by kyhavlov** [comment]
@chanwit the docker client throws a parse error when given that sort of syntax, specifically: `Error parsing reference: "nginx@constraint:node==node1,constraint:node!=node1" is not a valid repository/tag`.

Should I modify the docker CLI in some way to support this? The integration tests here use the CLI as well, so it seems necessary.

**Comment by schmunk42** [comment]
@justincormack is this fixed or just outdated?

**Comment by justincormack** [comment]
This was a pull request that had not been merged or commented on since 2016. As per https://github.com/docker/roadmap/issues/86 we are planning to deprecate Classic Swarm as, unlike Swarm/Swarmkit, no one is using it and it is not being maintained.

**Pull request closed without merging**